<a href="https://colab.research.google.com/github/rdhuwaraha21/DemoRepo1/blob/main/price_optimization_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##### Importing Required Libraries & Setting Up Matplotlib Inline for Plotting

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

##### Uploading CSV Using Google Colab File Upload



In [2]:
from google.colab import files
uploaded = files.upload()

Saving listings_full.csv to listings_full.csv


##### Reading CSV into a DataFrame

In [3]:
df = pd.read_csv("listings_full.csv")
df.head()

id                                      listing_url  \
0  6.624170e+06             https://www.airbnb.com/rooms/6624170   
1  8.837071e+06             https://www.airbnb.com/rooms/8837071   
2  7.161070e+05              https://www.airbnb.com/rooms/716107   
3  6.645390e+17  https://www.airbnb.com/rooms/664538756986273255   
4  8.191077e+06             https://www.airbnb.com/rooms/8191077   

      scrape_id last_scraped           source  \
0  2.024090e+13     9/5/2024  previous scrape   
1  2.024090e+13     9/5/2024  previous scrape   
2  2.024090e+13     9/5/2024  previous scrape   
3  2.024090e+13     9/6/2024  previous scrape   
4  2.024090e+13     9/5/2024  previous scrape   

                                             name  \
0             Warm, cozy sunlighted downtown appt   
1                   Cozy apartment in city center   
2              Loft style home nearby city centre   
3  Geweldige duurzame eco woonark op unieke plek!   
4                               Old bar apartment   

                                         description  \
0  2 room appt. 1.8 km from central station with ...   
1  Located in Amsterdam's sweet spot. A stone's t...   
2                                                NaN   
3  This unique eco houseboat is located in the mo...   
4  This just renovated apartment for 6 persons is...   

                               neighborhood_overview  \
0                                                NaN   
1                                      see the guide   
2                                                NaN   
3                                                NaN   
4  Our apartment is situated in the centre of Ams...   

                                         picture_url   host_id  ...  \
0  https://a0.muscache.com/pictures/df91da10-f7d4...  34670170  ...   
1  https://a0.muscache.com/pictures/5fee12d4-61d0...  13034277  ...   
2  https://a0.muscache.com/pictures/9927048/b367a...   3689867  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...   7647768  ...   
4  https://a0.muscache.com/pictures/miso/Hosting-...  43205475  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        4.86                   4.56                4.65   
1                         NaN                    NaN                 NaN   
2                        4.83                   4.66                4.61   
3                        5.00                   5.00                5.00   
4                        4.79                   4.83                4.45   

                    license instant_bookable calculated_host_listings_count  \
0  0363 3CF3 3233 3B24 6BEC                f                              1   
1  0363 0C20 1768 FAAA 3556                f                              1   
2  0363 0014 1E57 F097 81A5                f                              1   
3  0363 468D A0A1 6595 5E39                f                              1   
4                    Exempt                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           1   
1                                           1   
2                                           1   
3                                           1   
4                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.38  
1                                           0               NaN  
2                                           0              1.21  
3                                           0              0.04  
4                              

In [4]:
pd.set_option('display.max_rows', None)   # To view full dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 75 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            9659 non-null   float64
 1   listing_url                                   9659 non-null   object 
 2   scrape_id                                     9659 non-null   float64
 3   last_scraped                                  9659 non-null   object 
 4   source                                        9659 non-null   object 
 5   name                                          9659 non-null   object 
 6   description                                   9310 non-null   object 
 7   neighborhood_overview                         5469 non-null   object 
 8   picture_url                                   9659 non-null   object 
 9   host_id                                       9659 non-null   i

##### Selecting the Columns to have

In [5]:
pd.set_option('display.max_columns', None)
columns_to_keep = [
    'neighbourhood_cleansed', 'latitude', 'longitude', 'property_type', 'room_type', 'bedrooms',
    'accommodates', 'bathrooms_text', 'beds','amenities', 'host_is_superhost', 'host_identity_verified',
    'host_total_listings_count', 'availability_30', 'availability_60', 'availability_90', 'availability_365',
    'number_of_reviews', 'review_scores_rating','review_scores_accuracy',
    'review_scores_cleanliness', 'review_scores_checkin', 'review_scores_communication', 'review_scores_location',
    'review_scores_value', 'reviews_per_month', 'instant_bookable','minimum_nights', 'maximum_nights', 'price']
df_copy = df[columns_to_keep]
df_copy.head(4)


neighbourhood_cleansed  latitude  longitude       property_type  \
0            Centrum-Oost  52.36842    4.92438  Entire rental unit   
1  De Baarsjes - Oud-West  52.36455    4.87712  Entire rental unit   
2  De Baarsjes - Oud-West  52.36462    4.86824  Entire rental unit   
3               Oud-Noord  52.39394    4.90587           Houseboat   

         room_type  bedrooms  accommodates bathrooms_text  beds  \
0  Entire home/apt       1.0             3         1 bath   NaN   
1  Entire home/apt       1.0             2         1 bath   NaN   
2  Entire home/apt       1.0             2         1 bath   NaN   
3  Entire home/apt       2.0             3         1 bath   NaN   

                                           amenities host_is_superhost  \
0  ["Dishes and silverware", "Smoke alarm", "Hang...                 f   
1  ["Smoke alarm", "Essentials", "Iron", "Washer"...                 f   
2  ["Smoke alarm", "Essentials", "Iron", "Host gr...                 f   
3  ["Paid parking on premises", "Smoke alarm", "C...                 f   

  host_identity_verified  host_total_listings_count  availability_30  \
0                      t                          2                0   
1                      f                          1                0   
2                      t                          1                0   
3                      t                          4                0   

   availability_60  availability_90  availability_365  number_of_reviews  \
0                0                0                 0                 43   
1                0                0                 0                  0   
2                0                0                 1                172   
3                0                0                 0                  1   

   review_scores_rating  review_scores_accuracy  review_scores_cleanliness  \
0                  4.84                    4.83                       4.86   
1                   NaN                     NaN                        NaN   
2                  4.71                    4.82                       4.64   
3                  5.00                    5.00                       5.00   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                   4.83                         4.86                    4.56   
1                    NaN                          NaN                     NaN   
2                   4.78                         4.83                    4.66   
3                   5.00                         5.00                    5.00   

   review_scores_value  reviews_per_month instant_bookable  minimum_nights  \
0                 4.65               0.38                f               2   
1                  NaN                NaN                f               2   
2                 4.61               1.21                f               3   
3                 5.00               0.04                f               3   

   maximum_nights price  
0            1125   NaN  
1            1125   NaN  
2            1125   NaN  
3              21   NaN

#### Data Inspection

In [ ]:
df_copy.shape

##### Summary of Numerical Columns

In [ ]:
df_copy.describe()

##### Count of Missing Values

In [ ]:
df_copy.isnull().sum()

#### Duplicate Data Handling

In [ ]:
df_copy[df_copy.duplicated()]

In [ ]:
df_copy = df_copy.drop_duplicates(keep = "first")

##### Feature Type Identification

In [ ]:
numeric_features = [feature for feature in df_copy.columns if df_copy[feature].dtype != 'O']
categorical_features = [feature for feature in df_copy.columns if df_copy[feature].dtype == 'O']


print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

##### Analyzing Unique Values in Categorical Columns



In [ ]:
for col in categorical_features:
    unique_values = df_copy[col].unique()
    count_unique = df_copy[col].nunique()
    print(f"Unique values in '{col}': {unique_values}")
    print(f"Number of unique values in '{col}': {count_unique}\n")

# Handling Missing Values & Outliers

## Price
- Handling missing values and dtype changing

In this dataset, the price column is missing for over 3,000 rows out of a total of 9,659. When building a predictive model, it's generally better to remove rows with missing target values like price. Keeping them could lead to complications and inaccuracies during model training.

In [ ]:
df_copy = df_copy.dropna(subset=['price'])
print(f"Number of rows after dropping NaN values in 'price': {len(df_copy)}")

In [ ]:
df_copy.isnull().sum()

The price values were stored as strings with $ and , symbols, making them unusable for calculations or modeling. We cleaned and converted them to float type for numerical analysis.

In [ ]:
#$ sign and , replacing
df_copy['price'] = df_copy['price'].str.replace('$', '').str.replace(',', '') .str.strip().astype(float)
print(df_copy['price'].head())

In [ ]:
numeric_features = [feature for feature in df_copy.columns if df_copy[feature].dtype != 'O']
categorical_features = [feature for feature in df_copy.columns if df_copy[feature].dtype == 'O']

# print columns
print('We have {} numerical features'.format(len(numeric_features)))
print('\nWe have {} categorical features'.format(len(categorical_features)))

Below step gives us insight into the distribution and frequency of different prices. It helps identify outliers and unrealistic values that might affect the analysis or model performance.

In [ ]:
price_counts = df_copy['price'].value_counts()

price_counts_df = price_counts.reset_index()
price_counts_df.columns = ['Price', 'Count']

price_counts_sorted = price_counts_df.sort_values(by='Price', ascending=False)

print(price_counts_sorted)

To reduce the impact of outlier prices (e.g., listings above $3000), we applied a cutoff. These extreme values can skew the distribution and reduce model accuracy, especially in price prediction tasks.

In [ ]:
threshold = 1000
df_copy = df_copy[df_copy['price'] <= threshold]
df_copy.head()

#### Visualizing Cleaned Price Distribution

##### 1. Histogram – To See Overall Price Distribution

In [ ]:
plt.figure(figsize=(10, 5))
sns.histplot(df_copy['price'], bins=50, kde=True)
plt.title("Price Distribution (After Cleaning)", fontsize=14)
plt.xlabel("Price")
plt.ylabel("Count")
plt.show()


- The histogram shows that the majority of Airbnb listings are priced between 50 and 300.
-The distribution is right-skewed, meaning a small number of listings have much higher prices (closer to 1000 or even up to 3000).
-This confirms that while most listings are affordable, some luxury or premium listings still exist in the dataset.
-The 3000 threshold helped reduce extreme outliers, but a few mild outliers are still visible on the far right, which is typical in real-world pricing data.
-The peak around $150–$250 indicates the most common price range.

##### 2. Boxplot – To Detect Remaining Outliers Visually

In [ ]:
plt.figure(figsize=(10, 2))
sns.boxplot(x=df_copy['price'])
plt.title("Boxplot of Price (After Cleaning)", fontsize=14)
plt.xlabel("Price")
plt.show()


- The boxplot reveals that while most Airbnb listings fall below 500, there are many outliers beyond this range, extending up to 3000.
- The box represents the interquartile range (IQR)—the middle 50% of listings—mostly between 100 and 300.
- The horizontal line inside the box is the median price, showing where half the listings are cheaper and half are more expensive.
- The dots beyond the whiskers are outliers, indicating higher-priced listings that are much less frequent.
- Even after applying the $3000 threshold, the data remains right-skewed, which is typical in real-world pricing.

## Review scores

When a group of related columns, like the review_scores_*, have the same missing entries, it's crucial to evaluate whether the missingness is systematic (e.g., some hosts never received any reviews) or random. Here's how to handle this:

In [ ]:
print(df_copy.columns)


#####  1: Understanding the Pattern of Missing Review Scores

In [ ]:
print(df_copy[df_copy['number_of_reviews'] == 0]['review_scores_rating'].isnull().all())


- it confirms that all missing values in the review_scores_* columns are systematically tied to listings with number_of_reviews == 0. These are likely listings that have never received any reviews.
- Since these rows represent listings with no reviews, you can impute the review_scores_* columns with a default value, such as 0, to explicitly indicate "no reviews available." This prevents these rows from being excluded while ensuring the model interprets them correctly.
- The missing values in reviews_per_month are systematically tied to listings where number_of_reviews is 0. This makes sense because if a listing has no reviews, it cannot have a valid reviews_per_month value.

##### 2: Imputing Missing Review Scores with Median

In [ ]:
# Define columns for imputation
review_score_columns = [
    'review_scores_rating', 'review_scores_accuracy', 'review_scores_cleanliness',
    'review_scores_checkin', 'review_scores_communication', 'review_scores_location',
    'review_scores_value', 'reviews_per_month'
]

# Impute missing values with median
for col in review_score_columns:
    df_copy[col] = df_copy[col].fillna(df_copy[col].median())

# Verify that missing values are filled
print(df_copy[review_score_columns].isnull().sum())


- We used the median to fill missing values for all review score-related columns.
- Median is a robust measure that handles skewed data better than the mean.
- This approach keeps those rows usable for analysis while ensuring models interpret them correctly.
- It’s especially helpful when modeling, as dropping these rows would result in significant data loss.

In [ ]:
df_copy.isnull().sum()  # Verification

# Imputation and Distribution Analysis for Remaining Features

## Imputation and Distribution Analysis for Remaining Features

- finding the relationship between beds and other features like accommodates, room_type, and bedrooms is a sound approach before imputing missing values. Correlation analysis is one part of this process, but you might also need grouping and analysis to better understand patterns in the data.

#### 1: Imputing Missing Values in 'beds' Using Correlation-Based Grouping



In [ ]:
correlation = df_copy[['beds', 'accommodates', 'bedrooms']].corr()
print(correlation)

- A strong positive correlation between beds and accommodates (0.78).
- A moderate to strong positive correlation between beds and bedrooms (0.69).
- A strong positive correlation between accommodates and bedrooms (0.75).

In [ ]:
# Fill missing values in 'beds' based on median of groups
df_copy['beds'] = df_copy['beds'].fillna(df_copy.groupby(['accommodates', 'bedrooms'])['beds'].transform('median'))


1. Bedrooms (Numerical):
- Since bedrooms is a numeric variable, you can use a central tendency measure like the median to impute missing values. The median is less affected by outliers and provides a stable central value.
2. Bathrooms_text (Categorical):
- The bathrooms_text column is categorical. Usually, you can impute missing categorical values with the mode (the most frequent category).
If possible, consider converting bathrooms_text into a numerical feature (e.g., extracting the number of bathrooms) before imputation. Otherwise, just use the mode.
3. Host_is_superhost (Binary):
- host_is_superhost is a binary column (often t/f or True/False).
For binary columns, the most common approach is to fill missing values with the mode. If the majority of hosts are not superhosts, the mode is likely False (or f). This maintains the most common status in your data.

#### 2: Filling Other Columns Based on Data Type

In [ ]:
# bedrooms (numeric)
df_copy['bedrooms'] = df_copy['bedrooms'].fillna(df_copy['bedrooms'].median())

# bathrooms_text (categorical)
df_copy['bathrooms_text'] = df_copy['bathrooms_text'].fillna(df_copy['bathrooms_text'].mode()[0])

# host_is_superhost (binary)
df_copy['host_is_superhost'] = df_copy['host_is_superhost'].fillna(df_copy['host_is_superhost'].mode()[0])


In [ ]:
df_copy.isnull().sum()   # Verify

## Distribution Analysis of Categorical Features

In [ ]:
# Exclude 'amenities' from categorical_features since it has >9000
categorical_features = [col for col in categorical_features if col != 'amenities']

for col in categorical_features:
    print(f"Proportions for column '{col}':")
    print(df_copy[col].value_counts(normalize=True) * 100)
    print('---------------------------')


- The 'amenities' column was excluded from this proportion analysis due to its high cardinality and complex string-based format, which requires separate preprocessing (e.g., parsing and vectorization).


`neighbourhood_cleansed`:

The top 3 neighborhoods—De Baarsjes - Oud-West, Centrum-West, and De Pijp - Rivierenbuurt—account for over 40% of all listings.

`property_type`:

Over 38% of listings are Entire rental unit. Top 3 types cover a large share, while many niche types occur < 1% — these may be grouped as "Other" if needed.

`room_type`:

Majority (~75%) are Entire home/apt. Private room is next most common. This distribution may impact pricing and review scores significantly.

`bathrooms_text`:

Highly diverse. Most listings (~49%) have 1 bath, followed by 1.5 baths and 1 private bath. Consider simplifying this column (e.g., converting to number of bathrooms).

``host_is_superhost``:

About 77% of hosts are not superhosts (f), and 23% are. Useful as a binary classification feature.

``host_identity_verified``:

Almost all hosts (~98%) are verified. Since this is highly imbalanced, it may have low predictive power unless you're modeling trust.

``instant_bookable``:

About 81% are not instant bookable. This may reflect host behavior or listing type, and can influence guest booking preferences.

### Univariate Distribution Analysis of Numerical Features

In [ ]:
## Proportion of count data on numerical columns
import math

num_features = len(numeric_features)
cols = 3
rows = math.ceil(num_features / cols)

plt.figure(figsize=(15, rows * 4))
plt.suptitle('Univariate Analysis of Numerical Features', fontsize=20, fontweight='bold', alpha=0.8, y=1.0)

for i, feature in enumerate(numeric_features):
    plt.subplot(rows, cols, i + 1)
    sns.kdeplot(x=df_copy[feature], fill=True, color='r')
    plt.xlabel(feature)

plt.tight_layout(rect=[0, 0, 1, 0.96])
plt.show()


####The KDE plots reveal various distribution patterns in the numerical features:

- Location features (``latitude``, ``longitude``) show bell-shaped (normal) distributions centered around Amsterdam coordinates.

- Room characteristics (``bedrooms``, ``accommodates``, ``beds``) are highly right-skewed, with most listings offering smaller units.

- Host features (``host_total_listings_count``) are heavily skewed, indicating a few hosts manage many properties.

- Availability fields (``availability_30``, ``60``, ``90``, ``365``) show multimodal or exponential-like patterns.

- Review score features like review_scores_rating, accuracy, cleanliness, etc., are left-skewed (``clustered near 5``), showing most users leave high ratings.

- Engagement metrics (``reviews_per_month``, ``number_of_reviews``) are extremely right-skewed, with most listings receiving few or no reviews.

- Booking restrictions (minimum_nights, maximum_nights) also show wide dispersion and right skewness, hinting at some extreme outliers.

#### These insights help determine:

- Whether transformations (e.g., log, Box-Cox) are needed.

- Which features might need outlier treatment or binning.

- Which ones are already normally distributed and ready for modeling.

## Handling and Encoding Amenities Data

#### Transforming the complex string list of amenities into structured binary features using the top 20 most common amenities.

In [ ]:
import ast
from collections import Counter

df_copy['amenities'] = df_copy['amenities'].apply(lambda x: ast.literal_eval(x))

all_amenities = [item for sublist in df_copy['amenities'] for item in sublist]
top_amenities = [amenity for amenity, count in Counter(all_amenities).most_common(20)]  # Top 10 amenities

for amenity in top_amenities:
    df_copy[f'has_{amenity}'] = df_copy['amenities'].apply(lambda x: 1 if amenity in x else 0)

df_copy.drop('amenities', axis=1, inplace=True)

df_copy.head(3)


#### The ``amenities`` column originally contains a list of items stored as stringified Python lists (e.g.``Wifi``, ``TV``, ``Kitchen``), which is unsuitable for modeling in raw form.

- Step 1: Parsed string to list using ``ast.literal_eval()`` to safely convert the string into an actual list.

- Step 2: Counted all unique amenities using ``Counter`` and extracted the top 20 most common amenities.

- Step 3: Created new binary features for each of these top amenities. Each new column indicates whether a listing has a specific amenity (1) or not (0).

- Step 4: Dropped the original ``amenities`` column after encoding.

 This approach keeps the most relevant information without overloading the dataset with too many sparse features, and it prepares the data for machine learning models which require numerical input.

In [ ]:
amenity_counts = Counter(all_amenities).most_common(20)
amenity_names, counts = zip(*amenity_counts)

plt.figure(figsize=(12, 6))
plt.barh(amenity_names[::-1], counts[::-1], color='skyblue')
plt.title('Top 20 Most Common Amenities', fontsize=16, fontweight='bold')
plt.xlabel('Frequency')
plt.ylabel('Amenity')
plt.tight_layout()
plt.show()

#### This bar chart shows the top 20 amenities most frequently offered by hosts.

- Safety amenities like Smoke alarm and Carbon monoxide alarm are highly common, indicating compliance with guest safety expectations.

- Wifi, Hot water, Kitchen, and Essentials are also dominant, suggesting these are considered must-have basics.

- Dishes and silverware, Hair dryer, and Shampoo reflect hospitality focus on comfort and convenience.

This insight is valuable for both:

- Hosts aiming to improve their offering.

- Modeling, to decide which binary amenities to retain for prediction tasks.

## Handling Binary & Categorical Columns

#### Binary Encoding
Converting 't' and 'f' to 1 and 0

In [ ]:
# Binary encoding: Convert 't' and 'f' to 1 and 0
binary_columns = ['host_is_superhost', 'host_identity_verified', 'instant_bookable']

for col in binary_columns:
    df_copy[col] = df_copy[col].map({'t': 1, 'f': 0})

df_copy[binary_columns].head(3)


#### Identifying Categorical Columns
 Overview of Object-Type Columns

In [ ]:
categorical_columns = [col for col in df_copy.columns if df_copy[col].dtype == 'O']

print("Categorical Columns:", categorical_columns)


Counting Unique Values

In [ ]:
columns_to_select = ['neighbourhood_cleansed', 'property_type', 'room_type', 'bathrooms_text']
selected_data = df_copy[columns_to_select]

unique_count = selected_data.nunique()
print("Number of unique values in each selected column:\n", unique_count)


## Label Encoding of Categorical Columns

In [ ]:
from sklearn.preprocessing import LabelEncoder

columns_to_encode = ['neighbourhood_cleansed', 'property_type', 'room_type', 'bathrooms_text']

le = LabelEncoder()
for col in columns_to_encode:
    df_copy[col] = le.fit_transform(df_copy[col])

df_copy[columns_to_encode].head()


- In this step, label encoding is applied to convert categorical string values into numeric labels. This is essential for machine learning models that cannot handle non-numeric data directly.

- The columns encoded are: ``neighbourhood_cleansed``, ``property_type``, ``room_type``, and ``bathrooms_text``.

- Each unique category in these columns is replaced with a distinct integer using ``LabelEncoder``.

- This encoding retains the uniqueness of each category while making the data model-friendly.

In [ ]:
df_copy.head()


In [ ]:
df_copy.isnull().sum()

After performing a comprehensive data cleaning process — including handling missing values, encoding categorical and binary variables, and processing complex columns like amenities — we verified the dataset for any remaining missing values.

- All columns now have zero missing values.
This confirms the dataset is now fully clean and ready for modeling

### Visualizations

1. Correlation Heatmap

In [ ]:
plt.figure(figsize=(12,10))
sns.heatmap(df_copy.corr(), cmap='coolwarm', annot=False)
plt.title("Correlation Heatmap")
plt.show()

This heatmap helps identify relationships between all numerical features.

- Strong correlations (closer to 1 or -1) suggest linear relationships.

- Helps detect multicollinearity among features like ``review_scores_``*, ``availability_``*, etc.

- Useful for selecting or dropping features before model training.

2.  Feature vs Price Analysis

In [ ]:
sns.scatterplot(x='accommodates', y='price', data=df_copy)
plt.title('Accommodates vs Price')
plt.show()

This scatter plot reveals the relationship between the number of guests a property accommodates and its price.

- A general upward trend shows that price tends to increase with accommodates.

- However, there’s high variability and many outliers at each level, indicating the presence of other influencing factors.

 3. Categorical vs Price

In [ ]:
sns.boxplot(x='room_type', y='price', data=df_copy)
plt.xticks(rotation=45)
plt.title('Room Type vs Price')
plt.show()

This boxplot compares the price distribution across different encoded room types.
- ``Room_type = 0`` (likely "Entire home/apt") shows the highest price range and most variability.

- ``Room_type = 3`` (likely "Shared room") shows lower prices.

- Helpful in identifying pricing patterns based on room category.

# Model training

In [ ]:
from sklearn.model_selection import train_test_split

#  independent (X) and dependent (y) variables
X = df_copy.drop(['price'], axis=1)
y = df_copy['price']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Verify the splits
print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)


In [ ]:
X_train

- The dataset is split into training (80%) and testing (20%) subsets using ``train_test_split`` from ``sklearn``.

- This helps evaluate model performance on unseen data and reduces the risk of overfitting.

- After removing the target column (``price``) from X, both feature sets ``(X_train, X_test)`` and target sets ``(y_train, y_test)`` are properly aligned and ready for modeling.

### Model Training And Model Selection

## Model Selection and Evaluation Setup

To train and evaluate different regression models for predicting the price, we first need to import the required libraries and models. These include both ensemble and linear models, as well as metrics to measure performance like MAE, RMSE, and R².


In [ ]:
import pandas as pd
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from xgboost import XGBRegressor


- **Mean Absolute Error (MAE)** – average of absolute differences between predictions and actual values.
- **Root Mean Squared Error (RMSE)** – square root of the average squared differences.
- **R² Score** – indicates how well the model explains the variability of the target variable.


Define Evaluation Metrics

In [ ]:
##Create a Function to Evaluate Model
def evaluate_model(true, predicted):
    mae = mean_absolute_error(true, predicted)
    mse = mean_squared_error(true, predicted)
    rmse = np.sqrt(mean_squared_error(true, predicted))
    r2_square = r2_score(true, predicted)
    return mae, rmse, r2_square

This function will be used after each model makes predictions, allowing us to consistently measure and compare performance on both training and testing datasets.


### Begin Model Training and Evaluation

Define a dictionary of regression models to compare their performance. Each model will be:
- Trained using the training dataset
- Evaluated on both training and test datasets using the `evaluate_model` function
This setup enables quick benchmarking of different models using consistent metrics.


In [ ]:
## Beginning Model Training
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest Regressor": RandomForestRegressor(),
    "Adaboost Regressor":AdaBoostRegressor(),
     "XGBoost Regressor": XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)}

for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train, y_train) # Train model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate Train and Test dataset
    model_train_mae , model_train_rmse, model_train_r2 = evaluate_model(y_train, y_train_pred)

    model_test_mae , model_test_rmse, model_test_r2 = evaluate_model(y_test, y_test_pred)


    print(list(models.keys())[i])

    print('Model performance for Training set')
    print("- Root Mean Squared Error: {:.4f}".format(model_train_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_train_mae))
    print("- R2 Score: {:.4f}".format(model_train_r2))

    print('----------------------------------')

    print('Model performance for Test set')
    print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
    print("- R2 Score: {:.4f}".format(model_test_r2))

    print('='*35)
    print('\n')

### 📊 Model Performance Summary

| Model                   | Train RMSE | Train MAE | Train R² | Test RMSE | Test MAE | Test R² |
|------------------------|------------|-----------|----------|-----------|----------|---------|
| **Linear Regression**   | 136.50     | 80.12     | 0.4728   | 151.76    | 84.19    | 0.3951  |
| **Decision Tree**       | 0.39       | 0.01      | 1.0000   | 195.16    | 104.80   | -0.0005 |
| **Random Forest**       | 47.57      | 26.50     | 0.9360   | 140.23    | 74.80    | 0.4835  |
| **Adaboost**            | 237.31     | 213.15    | -0.5936  | 258.62    | 223.01   | -0.7568 |
| **XGBoost**             | 103.86     | 60.38     | 0.6948   | 137.85    | 73.21    | 0.5009  |


**Note:**  
### ✅ Final Model Selection

Based on the test set performance, the **XGBoost Regressor** was selected as the final model for this project. It achieved the best balance between bias and variance with:

- **R² Score:** 0.5009
- **RMSE:** 137.85
- **MAE:** 73.21

While the Random Forest Regressor also performed well (R² = 0.4835), XGBoost showed slightly better generalization. Other models such as Decision Tree and Adaboost were either highly overfitted or underperforming.

This final model will be used for generating insights, interpreting feature importance, and building a business-facing Power BI dashboard.



## Feature Importance – XGBoost

In [ ]:
# Assuming your trained XGBoost model is stored as:
xgb_model = models['XGBoost Regressor']

# Get feature importances
importances = xgb_model.feature_importances_
features = X.columns


In [ ]:
# Create DataFrame
xgb_feature_df = pd.DataFrame({'Feature': features, 'Importance': importances})
xgb_feature_df = xgb_feature_df.sort_values(by='Importance', ascending=False)

Plot the Top N Important Features

In [ ]:
# Plot Top 15
plt.figure(figsize=(10, 6))
sns.barplot(x='Importance', y='Feature', data=xgb_feature_df.head(15), palette='viridis')
plt.title('Top 15 Most Important Features (XGBoost)', fontsize=14)
plt.xlabel('Feature Importance Score')
plt.ylabel('Feature')
plt.tight_layout()
plt.show()

### Feature Importance Analysis (XGBoost)

The chart below presents the top 15 features contributing most to price prediction, based on the XGBoost model.

- **Top Features**:
  - `bedrooms`, `accommodates`, and `availability_60` are the most influential features.
  - Other key features include `room_type`, `bathrooms_text`, and amenities like `has_Dishwasher`.




##  XGBoost Tuning Code with ``RandomizedSearchCV``:

Even though XGBoost is a powerful algorithm, its default settings are not always optimal for every dataset.
Hyperparameter tuning helps to:

- Improve Prediction Accuracy
- Reduce Overfitting or Underfitting
- Balance model complexity and generalization
- Find the best configuration automatically using cross-validation

In [ ]:
from sklearn.model_selection import RandomizedSearchCV
from xgboost import XGBRegressor

# Define the model
xgb = XGBRegressor(random_state=42)

# Define the hyperparameter space
param_dist = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0],
    'gamma': [0, 1, 5]
}

# Set up RandomizedSearchCV
random_search = RandomizedSearchCV(
    estimator=xgb,
    param_distributions=param_dist,
    n_iter=20,  # Try 20 random combinations
    scoring='neg_mean_squared_error',
    cv=3,
    verbose=1,
    random_state=42,
    n_jobs=-1
)

# Fit the model
random_search.fit(X_train, y_train)

# Best parameters
print("Best Hyperparameters for XGBoost:")
print(random_search.best_params_)

 Best XGBoost Parameters:
- n_estimators       = 200
- max_depth          = 3
- learning_rate      = 0.1
- subsample          = 0.8
- colsample_bytree   = 0.8
- gamma              = 1


This code automates the process of finding the best XGBoost hyperparameters (like learning rate, depth, gamma, etc.) using cross-validation, improving the model's ability to generalize well to unseen data.


### Retraining XGBoost with Best Hyperparameters

In [ ]:
# Best hyperparameters found from RandomizedSearchCV
best_params = {
    'subsample': 0.8,
    'n_estimators': 200,
    'max_depth': 3,
    'learning_rate': 0.1,
    'gamma': 1,
    'colsample_bytree': 0.8
}

# Retrain XGBoost with best params
from xgboost import XGBRegressor

best_xgb = XGBRegressor(**best_params, random_state=42)
best_xgb.fit(X_train, y_train)

# Make predictions
y_train_pred = best_xgb.predict(X_train)
y_test_pred = best_xgb.predict(X_test)

# Evaluate using your existing function
train_mae, train_rmse, train_r2 = evaluate_model(y_train, y_train_pred)
test_mae, test_rmse, test_r2 = evaluate_model(y_test, y_test_pred)

# Print results
print("Tuned XGBoost Regressor Performance")
print(f"Train RMSE: {train_rmse:.2f}, MAE: {train_mae:.2f}, R²: {train_r2:.4f}")
print(f"Test  RMSE: {test_rmse:.2f}, MAE: {test_mae:.2f}, R²: {test_r2:.4f}")




After applying hyperparameter tuning, the XGBoost Regressor showed a noticeable improvement in test performance:

| Metric     | Train Set | Test Set |
|------------|-----------|----------|
| RMSE       | 94.04     | 137.59   |
| MAE        | 55.55     | 70.97    |
| R² Score   | 0.7498    | 0.5027   |

 **Key Insight**:
- The **R² score improved to 0.50** on the test set, the best among all models tested.
- The gap between training and test error is acceptable, showing **good generalization**.
- This tuned XGBoost model is selected as the **final model** for insights and deployment.


In [ ]:
import joblib

# Save the tuned XGBoost model to a file
joblib.dump(best_xgb, 'tuned_xgboost_model.pkl')
print("Model saved as 'tuned_xgboost_model.pkl'")


### Visualization of Residuals (Prediction Errors)

In [ ]:
# Add predictions and errors to a DataFrame
result_df = X_test.copy()
result_df['actual_price'] = y_test
result_df['predicted_price'] = y_test_pred
result_df['error'] = result_df['predicted_price'] - result_df['actual_price']
result_df['abs_error'] = result_df['error'].abs()

# Preview top rows with biggest absolute errors
result_df.sort_values(by='abs_error', ascending=False).head(10)


In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

residuals = y_test - y_test_pred

plt.figure(figsize=(8, 5))
sns.histplot(residuals, kde=True, bins=40)
plt.title("Distribution of Residuals (Prediction Error)")
plt.xlabel("Residual (Actual - Predicted Price)")
plt.ylabel("Frequency")
plt.axvline(0, color='red', linestyle='--')
plt.tight_layout()
plt.show()


In [ ]:
plt.figure(figsize=(8, 5))
sns.scatterplot(x=y_test_pred, y=residuals)
plt.axhline(0, color='red', linestyle='--')
plt.title("Residuals vs Predicted Prices")
plt.xlabel("Predicted Price")
plt.ylabel("Residual (Error)")
plt.tight_layout()
plt.show()


### Residual Analysis

Residuals help us understand how far off the predictions are from actual values. A good model should have residuals:

- Centered around zero
- Randomly scattered without patterns
- With consistent spread across all prediction levels

**Findings:**
- The residual histogram shows a near-normal distribution centered at zero, indicating low bias.
- The residual scatterplot shows increased error variance for higher price predictions. This could be explored further but remains acceptable for many real-world scenarios.

Overall, residuals confirm that the tuned XGBoost model performs reasonably well, especially on mid-range predictions.


### Segment-Based Prediction Error Analysis (Post Model Training)

Even if the overall model performance looks acceptable, it might underperform in certain price ranges, which could be critical in a business context—especially in markets where pricing precision drives customer decisions or revenue.

This analysis helps us:

- Identify weak spots where the model makes large errors (e.g., low or high price ranges).

- Understand error consistency across different segments of the price variable.

- Provide targeted insights for future model improvements (e.g., separate models for luxury or budget listings).

- Justify the model's reliability in segments with higher volume or business impact.

By evaluating residuals (prediction errors) within price segments, we ensure that the model is not only statistically strong overall, but also fair and robust across different user groups or product categories.

In [ ]:
# Segments:
bins = [0, 100, 200, 300, 500, 700, 1000, 2000]
labels = ['<100', '100–200', '200–300', '300–500', '500–700', '700–1000', '1000–2000']


In [ ]:
# Create DataFrame with true, predicted and residuals
results_df = pd.DataFrame({
    'Actual': y_test,
    'Predicted': y_test_pred,
})
results_df['Residual'] = results_df['Actual'] - results_df['Predicted']

# Segment prices
results_df['Price_Segment'] = pd.cut(results_df['Actual'], bins=bins, labels=labels)

# Group by segment and calculate average error
segment_errors = results_df.groupby('Price_Segment').agg({
    'Residual': ['mean', 'median', 'std', 'count']
}).reset_index()

# Display results
import pandas as pd
print(segment_errors)




To ensure our model performs fairly across all price levels, we analyzed residuals (prediction errors) in different price bands.

####  Key Findings:
- For prices **below 300**, the model **underpredicts** slightly (by 20–40 on average), which is manageable in many business contexts.
- Between **300–500**, the model begins **overpredicting** modestly.
- For listings **above 500**, the model consistently and significantly **overestimates prices**.
  - For example, in the 700–1000 range, the model is off by nearly **300** on average.
  - For 1000–2000 listings, it's off by over **700**, making it **unreliable** for premium listings.



In [ ]:
# Prepare data for plotting
segment_plot = segment_errors.copy()
segment_plot.columns = ['Price_Segment', 'Mean_Residual', 'Median_Residual', 'Std_Residual', 'Count']

# Plot mean residuals by price segment
plt.figure(figsize=(10, 6))
sns.barplot(data=segment_plot, x='Price_Segment', y='Mean_Residual', palette='coolwarm')

plt.axhline(0, color='gray', linestyle='--')
plt.title('📊 Average Residuals by Price Segment', fontsize=14)
plt.xlabel('Price Segment')
plt.ylabel('Average Prediction Error (Residual)')
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()


This bar chart shows how our model’s prediction errors (residuals) vary across different price segments.

- Segments below 300 tend to be **underpredicted** (negative bars).
- Segments above 500 are **overpredicted**, and the error grows significantly.
- This pattern confirms that the model behaves differently in low vs. high price ranges — indicating a need for **price-based segmentation** or **model refinement**.


# Extended Analysis

### Segment-Wise Modeling

Due to large residuals in the 557–1000 price range, we trained two separate models:
- One for listings priced between 11 and 557
- Another for listings priced between 557 and 1000

This strategy ensures that each model learns **segment-specific patterns**, improving overall accuracy and reducing bias from outliers.

#### Reason for Price-Based Modeling
Segment-wise residual analysis revealed that:
- Errors below 557 were relatively stable.
- Errors above 557 increased sharply, indicating poor generalization.
Thus, we train separate models for low- and high-price segments to improve prediction accuracy.



In [ ]:
# Segment A: df_low
df_low = df_copy[df_copy['price'] <= 557].copy()
X_low = df_low.drop('price', axis=1)
y_low = df_low['price']

X_low_train, X_low_test, y_low_train, y_low_test = train_test_split(X_low, y_low, test_size=0.2, random_state=42)

print("Segment A Train:", X_low_train.shape)
print("Segment A Test:", X_low_test.shape)

# Segment B: df_high
df_high = df_copy[(df_copy['price'] > 557) & (df_copy['price'] <= 1000)].copy()
X_high = df_high.drop('price', axis=1)
y_high = df_high['price']

X_high_train, X_high_test, y_high_train, y_high_test = train_test_split(X_high, y_high, test_size=0.2, random_state=42)

print("Segment B Train:", X_high_train.shape)
print("Segment B Test:", X_high_test.shape)



**Segment-Based Splitting**

Instead of one unified train-test split, we create two sub-datasets based on the price threshold of 557. Each sub-dataset has its own train-test split to produce two dedicated models, one for lower-priced listings, and another for higher-priced listings. This approach addresses the varying error patterns observed in the residual analysis.


In [ ]:
from xgboost import XGBRegressor

# Segment A Model
model_low = XGBRegressor(n_estimators=100, random_state=42)  # or use best hyperparams
model_low.fit(X_low_train, y_low_train)

# Predictions
y_low_train_pred = model_low.predict(X_low_train)
y_low_test_pred = model_low.predict(X_low_test)

# Evaluate
def evaluate_model(true, pred):
    from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
    import numpy as np
    mae = mean_absolute_error(true, pred)
    rmse = np.sqrt(mean_squared_error(true, pred))
    r2 = r2_score(true, pred)
    return mae, rmse, r2

train_mae_low, train_rmse_low, train_r2_low = evaluate_model(y_low_train, y_low_train_pred)
test_mae_low, test_rmse_low, test_r2_low = evaluate_model(y_low_test, y_low_test_pred)

print("Segment A Model:")
print(f"Train -> RMSE: {train_rmse_low:.2f}, MAE: {train_mae_low:.2f}, R2: {train_r2_low:.4f}")
print(f"Test  -> RMSE: {test_rmse_low:.2f}, MAE: {test_mae_low:.2f}, R2: {test_r2_low:.4f}")
print("="*40)


## Segment A (Price ≤ 557) – Final Model Summary

This segment showed consistent and reliable performance using the **XGBoost Regressor**.

### Performance Metrics:
- **Train Set**:
  - RMSE: 18.17
  - MAE: 12.74
  - R²: 0.9715
- **Test Set**:
  - RMSE: 69.93
  - MAE: 52.25
  - R²: 0.5742

### Why Segment A Performed Well:
- **Large volume of data** (~4,500+ listings) gave the model enough patterns to learn from.
- **Lower variance in pricing** (more standardized listings).
- **Features like bedrooms, room_type, and availability** had strong influence and were well captured.

### Modeling Decision:
- Chose **XGBoost** for its balance between bias and variance.
- Didn’t show signs of overfitting.
- No further tuning needed beyond initial performance.




In [ ]:
# Segment B Model
model_high = XGBRegressor(n_estimators=100, random_state=42)  # or use best hyperparams
model_high.fit(X_high_train, y_high_train)

# Predictions
y_high_train_pred = model_high.predict(X_high_train)
y_high_test_pred = model_high.predict(X_high_test)

# Evaluate
train_mae_high, train_rmse_high, train_r2_high = evaluate_model(y_high_train, y_high_train_pred)
test_mae_high, test_rmse_high, test_r2_high = evaluate_model(y_high_test, y_high_test_pred)

print("Segment B Model:")
print(f"Train -> RMSE: {train_rmse_high:.2f}, MAE: {train_mae_high:.2f}, R2: {train_r2_high:.4f}")
print(f"Test  -> RMSE: {test_rmse_high:.2f}, MAE: {test_mae_high:.2f}, R2: {test_r2_high:.4f}")
print("="*40)


### Segment-Wise Model Evaluation Summary

After identifying performance issues in the overall model using segment-based residual analysis, we trained separate models for different price segments to improve prediction accuracy.

#### Segment A (Price ≤ 557)
- **Model Used**: XGBoost Regressor
- **Training Performance**:  
  - RMSE: 18.17 | MAE: 12.74 | R²: 0.9715  
- **Testing Performance**:  
  - RMSE: 69.93 | MAE: 52.25 | R²: 0.5742  
-  **Interpretation**: Good generalization. The model performs well on both train and test sets, capturing the majority of the variance in low-price listings.

#### Segment B (Price 557–1000)
- **Model Used**: XGBoost Regressor
- **Training Performance**:  
  - RMSE: 0.01 | MAE: 0.01 | R²: 1.0000  
- **Testing Performance**:  
  - RMSE: 127.21 | MAE: 96.07 | R²: -0.2624  
- **Interpretation**: Severe overfitting — the model performs perfectly on training data but fails to generalize on unseen data.

---

### Next Step: Fixing Segment B

The overfitting in Segment B is likely due to:
- **Small dataset size** (only ~260 records)
- **Complex model** (XGBoost is prone to overfitting in low-data scenarios)

 **Solution**:
We will replace XGBoost with a **regularized linear model** like **Ridge Regression** or **Lasso Regression** for this segment. These simpler models can generalize better when data is limited.

This will help:
- Reduce variance
- Improve generalization
- Maintain interpretability

We’ll compare performance again after this adjustment.

---


### Ridge Regression for Segment B



In [ ]:
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Define a pipeline to scale features and apply Ridge regression
ridge_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('ridge', Ridge(alpha=1.0, random_state=42))
])

# Train the Ridge model on Segment B
ridge_pipeline.fit(X_high_train, y_high_train)

# Predictions
y_high_train_pred = ridge_pipeline.predict(X_high_train)
y_high_test_pred = ridge_pipeline.predict(X_high_test)

# Evaluation
train_mae_ridge, train_rmse_ridge, train_r2_ridge = evaluate_model(y_high_train, y_high_train_pred)
test_mae_ridge, test_rmse_ridge, test_r2_ridge = evaluate_model(y_high_test, y_high_test_pred)

# Print results
print("Segment B Ridge Regression Model:")
print(f"Train -> RMSE: {train_rmse_ridge:.2f}, MAE: {train_mae_ridge:.2f}, R2: {train_r2_ridge:.4f}")
print(f"Test  -> RMSE: {test_rmse_ridge:.2f}, MAE: {test_mae_ridge:.2f}, R2: {test_r2_ridge:.4f}")
print("="*40)


- Train RMSE and R² show slight improvement compared to XGBoost (which completely overfit), but...

- Test R² is still negative, indicating poor generalization — the model performs worse than predicting the mean.

- This segment might be too small (only 264 rows) and too noisy for stable modeling using standard regressors.

### Lasso regression on Segment B

In [ ]:
from sklearn.linear_model import Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# Segment B: Lasso Regression
lasso_model = make_pipeline(StandardScaler(), Lasso(alpha=0.1, random_state=42))
lasso_model.fit(X_high_train, y_high_train)

# Predictions
y_high_train_pred = lasso_model.predict(X_high_train)
y_high_test_pred = lasso_model.predict(X_high_test)

# Evaluate
train_mae_lasso, train_rmse_lasso, train_r2_lasso = evaluate_model(y_high_train, y_high_train_pred)
test_mae_lasso, test_rmse_lasso, test_r2_lasso = evaluate_model(y_high_test, y_high_test_pred)

print("Segment B Lasso Regression Model:")
print(f"Train -> RMSE: {train_rmse_lasso:.2f}, MAE: {train_mae_lasso:.2f}, R2: {train_r2_lasso:.4f}")
print(f"Test  -> RMSE: {test_rmse_lasso:.2f}, MAE: {test_mae_lasso:.2f}, R2: {test_r2_lasso:.4f}")
print("="*40)

In [ ]:
from sklearn.linear_model import Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.ensemble import GradientBoostingRegressor

# Segment B: Gradient Boosting Regressor
gb_model = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)
gb_model.fit(X_high_train, y_high_train)

# Predictions
y_high_train_pred = gb_model.predict(X_high_train)
y_high_test_pred = gb_model.predict(X_high_test)

# Evaluate
train_mae_gb, train_rmse_gb, train_r2_gb = evaluate_model(y_high_train, y_high_train_pred)
test_mae_gb, test_rmse_gb, test_r2_gb = evaluate_model(y_high_test, y_high_test_pred)

print("Segment B Gradient Boosting Model:")
print(f"Train -> RMSE: {train_rmse_gb:.2f}, MAE: {train_mae_gb:.2f}, R2: {train_r2_gb:.4f}")
print(f"Test  -> RMSE: {test_rmse_gb:.2f}, MAE: {test_mae_gb:.2f}, R2: {test_r2_gb:.4f}")
print("="*40)

In [ ]:
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# Create a pipeline with scaling and SVR
svr_model = make_pipeline(StandardScaler(), SVR(kernel='rbf', C=100, gamma=0.1, epsilon=10))
svr_model.fit(X_high_train, y_high_train)

# Predict
y_high_train_pred = svr_model.predict(X_high_train)
y_high_test_pred = svr_model.predict(X_high_test)

# Evaluate
train_mae_svr, train_rmse_svr, train_r2_svr = evaluate_model(y_high_train, y_high_train_pred)
test_mae_svr, test_rmse_svr, test_r2_svr = evaluate_model(y_high_test, y_high_test_pred)

print("Segment B SVR Model:")
print(f"Train -> RMSE: {train_rmse_svr:.2f}, MAE: {train_mae_svr:.2f}, R2: {train_r2_svr:.4f}")
print(f"Test  -> RMSE: {test_rmse_svr:.2f}, MAE: {test_mae_svr:.2f}, R2: {test_r2_svr:.4f}")
print("="*40)


## Segment B (Price: 557–1000) – Model Performance Summary

Despite multiple attempts with advanced regression techniques, the model consistently struggled to generalize well on the test set for Segment B.

### Why Segment B Performed Poorly:
- **Limited Data**: Only ~264 listings were available in this segment, making it difficult for models to learn generalizable patterns.
- **High Price Volatility**: Listings in this range had large price fluctuations (e.g., due to luxury amenities or premium locations) which weren't fully explained by the available features.
- **Overfitting Risk**: Several models (like XGBoost and Gradient Boosting) performed extremely well on training data but failed on test data — indicating overfitting.
- **Feature Limitations**: Some influential factors (e.g., listing descriptions, seasonal pricing, photos) may be missing from the dataset.

### Models Tried in Segment B:
- XGBoost Regressor
- Ridge Regression
- Lasso Regression
- Gradient Boosting Regressor
- Support Vector Regressor (SVR)
- LightGBM Regressor

###  How We Handled It:
- We kept all model attempts to showcase **experimentation and transparency**.
- **Clearly reported** R², MAE, and RMSE for each model to help interpret where the gap lies.
- Acknowledged limitations in the data rather than artificially inflating results.
- Suggested that in real-world deployment, this segment may require:
  - More feature engineering or external data sources
  - Business rule-based heuristics (e.g., fallback price band estimation)
  - Specialized models trained with domain-specific attributes




In [ ]:
import joblib

# Save Segment A XGBoost model
joblib.dump(model_low, 'segment_a_xgboost_model.pkl')
print("Segment A XGBoost model saved successfully!")


In [ ]:
# Calculate residuals (Prediction Error)
residuals_low = y_low_test - y_low_test_pred

# Histogram of Residuals
plt.figure(figsize=(8, 5))
sns.histplot(residuals_low, bins=50, kde=True, color='skyblue')
plt.axvline(0, color='red', linestyle='--')
plt.title('Distribution of Residuals (Segment A)')
plt.xlabel('Residual (Actual - Predicted Price)')
plt.ylabel('Frequency')
plt.tight_layout()
plt.show()

# Residuals vs Predicted
plt.figure(figsize=(8, 5))
sns.scatterplot(x=y_low_test_pred, y=residuals_low)
plt.axhline(0, color='red', linestyle='--')
plt.title('Residuals vs Predicted Price (Segment A)')
plt.xlabel('Predicted Price')
plt.ylabel('Residual')
plt.tight_layout()
plt.show()

#### Residual Visualization (Segment A)
1. Distribution of Residuals
The residuals (actual - predicted prices) form a nearly normal distribution centered around zero.
This shape indicates that the model’s errors are mostly small and evenly distributed on both sides of the actual values.

  - Conclusion: The model is not biased toward underpredicting or overpredicting in this price segment.

2. Residuals vs Predicted Prices
The residuals are randomly dispersed around the horizontal zero line.
There is no clear pattern or funnel shape, which suggests that:
The variance of errors is stable across the prediction range (no heteroscedasticity).
The model is generalizing well for this segment.

### Getting the labels of catgorical columns

In [ ]:
# Proper mapping from original data (make sure to use unencoded columns)
from sklearn.preprocessing import LabelEncoder

# 🏘 Neighbourhood
le_neigh = LabelEncoder()
le_neigh.fit(df['neighbourhood_cleansed'])  # original column
neigh_map = {i: label for i, label in enumerate(le_neigh.classes_)}

# 🏡 Property Type
le_property = LabelEncoder()
le_property.fit(df['property_type'])
property_map = {i: label for i, label in enumerate(le_property.classes_)}

# 🛏 Room Type
le_room = LabelEncoder()
le_room.fit(df['room_type'])
room_map = {i: label for i, label in enumerate(le_room.classes_)}

# 🛁 Bathrooms Text
le_bath = LabelEncoder()
le_bath.fit(df['bathrooms_text'])
bath_map = {i: label for i, label in enumerate(le_bath.classes_)}

# ✅ Print human-readable mappings
print("🏘 Neighbourhood Mapping:", neigh_map)
print("🏡 Property Type Mapping:", property_map)
print("🛏 Room Type Mapping:", room_map)
print("🛁 Bathrooms Text Mapping:", bath_map)


In [ ]:
segment_a_results.columns


In [ ]:
import pandas as pd

# 1. Recreate segment_a_results
segment_a_results = X_low_test.copy()
segment_a_results['Actual Price ($)'] = y_low_test
segment_a_results['Predicted Price ($)'] = y_low_test_pred
segment_a_results['Prediction Error ($)'] = y_low_test - y_low_test_pred
segment_a_results.reset_index(drop=True, inplace=True)

In [ ]:
#  Neighbourhood Mapping
neighbourhood_mapping = {
    0: 'Bijlmer-Centrum',
    1: 'Bijlmer-Oost',
    2: 'Bos en Lommer',
    3: 'Buitenveldert - Zuidas',
    4: 'Centrum-Oost',
    5: 'Centrum-West',
    6: 'De Aker - Nieuw Sloten',
    7: 'De Baarsjes - Oud-West',
    8: 'De Pijp - Rivierenbuurt',
    9: 'Gaasperdam - Driemond',
    10: 'Geuzenveld - Slotermeer',
    11: 'IJburg - Zeeburgereiland',
    12: 'Noord-Oost',
    13: 'Noord-West',
    14: 'Oostelijk Havengebied - Indische Buurt',
    15: 'Osdorp',
    16: 'Oud-Noord',
    17: 'Oud-Oost',
    18: 'Slotervaart',
    19: 'Watergraafsmeer',
    20: 'Westerpark',
    21: 'Zuid'
}

# Apply decoding to the column
segment_a_results['Neighbourhood'] = segment_a_results['neighbourhood_cleansed'].map(neighbourhood_mapping)


In [ ]:
# Replace encoded property type values with actual names
property_type_mapping = {
    0: 'Barn', 1: 'Boat', 2: 'Camper/RV', 3: 'Casa particular', 4: 'Cave', 5: 'Earthen home',
    6: 'Entire cabin', 7: 'Entire chalet', 8: 'Entire condo', 9: 'Entire cottage', 10: 'Entire guest suite',
    11: 'Entire guesthouse', 12: 'Entire home', 13: 'Entire loft', 14: 'Entire place', 15: 'Entire rental unit',
    16: 'Entire serviced apartment', 17: 'Entire townhouse', 18: 'Entire vacation home', 19: 'Entire villa',
    20: 'Houseboat', 21: 'Private room', 22: 'Private room in barn', 23: 'Private room in bed and breakfast',
    24: 'Private room in boat', 25: 'Private room in bungalow', 26: 'Private room in cabin',
    27: 'Private room in casa particular', 28: 'Private room in condo', 29: 'Private room in earthen home',
    30: 'Private room in farm stay', 31: 'Private room in guest suite', 32: 'Private room in guesthouse',
    33: 'Private room in home', 34: 'Private room in hostel', 35: 'Private room in houseboat',
    36: 'Private room in hut', 37: 'Private room in loft', 38: 'Private room in nature lodge',
    39: 'Private room in rental unit', 40: 'Private room in serviced apartment', 41: 'Private room in tiny home',
    42: 'Private room in townhouse', 43: 'Private room in vacation home', 44: 'Private room in villa',
    45: 'Room in aparthotel', 46: 'Room in bed and breakfast', 47: 'Room in boutique hotel',
    48: 'Room in hostel', 49: 'Room in hotel', 50: 'Room in serviced apartment', 51: 'Shared room in boat',
    52: 'Shared room in home', 53: 'Shared room in hostel', 54: 'Shared room in hotel',
    55: 'Shared room in houseboat', 56: 'Shared room in rental unit', 57: 'Tiny home', 58: 'Yurt'
}

segment_a_results['Property Type'] = segment_a_results['property_type'].map(property_type_mapping)


In [ ]:
# Room Type mapping
room_type_mapping = {
    0: 'Entire home/apt',
    1: 'Hotel room',
    2: 'Private room',
    3: 'Shared room'
}

# Apply mapping
segment_a_results['Room Type'] = segment_a_results['room_type'].map(room_type_mapping)


In [ ]:
# Bathrooms (Text) mapping
bathroom_text_mapping = {
    0: '0 baths',
    1: '0 shared baths',
    2: '1 bath',
    3: '1 private bath',
    4: '1 shared bath',
    5: '1.5 baths',
    6: '1.5 shared baths',
    7: '12 baths',
    8: '17 baths',
    9: '2 baths',
    10: '2 shared baths',
    11: '2.5 baths',
    12: '2.5 shared baths',
    13: '3 baths',
    14: '3 shared baths',
    15: '3.5 baths',
    16: '4 baths',
    17: '4 shared baths',
    18: '4.5 baths',
    19: '5 baths',
    20: '5.5 baths',
    21: '7 baths',
    22: 'Half-bath',
    23: 'Private half-bath',
    24: 'Shared half-bath',
    25: 'Unknown'
}

# Apply mapping
segment_a_results['Bathrooms (Text)'] = segment_a_results['bathrooms_text'].map(bathroom_text_mapping)


In [ ]:
# Mapping dictionaries
binary_mapping = {'t': 'Yes', 'f': 'No'}

# Apply to relevant columns
segment_a_results['Superhost Status'] = segment_a_results['host_is_superhost'].map(binary_mapping)
segment_a_results['Host Identity Verified'] = segment_a_results['host_identity_verified'].map(binary_mapping)
segment_a_results['Instant Booking Available'] = segment_a_results['instant_bookable'].map(binary_mapping)


In [ ]:
# List of binary amenity columns
amenity_columns = [
    'has_Smoke alarm', 'has_Wifi', 'has_Hot water',
    'has_Kitchen', 'has_Essentials', 'has_Dishes and silverware',
    'has_Bed linens', 'has_Hair dryer', 'has_Refrigerator', 'has_Hangers',
    'has_Iron', 'has_Cooking basics', 'has_Carbon monoxide alarm',
    'has_Dedicated workspace', 'has_Hot water kettle', 'has_Shampoo',
    'has_Dishwasher', 'has_Wine glasses', 'has_Heating', 'has_Dining table'
]

# Replace 1 -> 'Yes' and 0 -> 'No' for each
for col in amenity_columns:
    segment_a_results[col] = segment_a_results[col].map({1: 'Yes', 0: 'No'})


In [ ]:
segment_a_results.columns


### Preparing the dataset for Power BI

In [ ]:
# Save to CSV
segment_a_results.to_csv('SegmentA_Predictions_updated.csv', index=False)


In [ ]:
from google.colab import files
files.download('SegmentA_Predictions_updated.csv')
